// @aboutcircles/miniapp-sdk/dist/index.js
// src/index.ts
var _address = null;
var _listeners = [];
var _dataListeners = [];
var _requestCounter = 0;
var _pending = {};
if (typeof window !== "undefined") {
  window.addEventListener("message", (event) => {
    const d = event.data;
    if (!d || !d.type)
      return;
    switch (d.type) {
      case "app_data":
        _dataListeners.forEach((fn) => fn(d.data));
        break;
      case "wallet_connected":
        _address = d.address;
        _listeners.forEach((fn) => fn(_address));
        break;
      case "wallet_disconnected":
        _address = null;
        _listeners.forEach((fn) => fn(null));
        break;
      case "tx_success":
        _pending[d.requestId]?.resolve(d.hashes);
        delete _pending[d.requestId];
        break;
      case "tx_rejected":
        _pending[d.requestId]?.reject(new Error(d.error ?? d.reason ?? "Rejected"));
        delete _pending[d.requestId];
        break;
      case "sign_success":
        _pending[d.requestId]?.resolve({
          signature: d.signature,
          verified: d.verified
        });
        delete _pending[d.requestId];
        break;
      case "sign_rejected":
        _pending[d.requestId]?.reject(new Error(d.error ?? d.reason ?? "Rejected"));
        delete _pending[d.requestId];
        break;
    }
  });
  if (window.parent !== window) {
    window.parent.postMessage({ type: "request_address" }, "*");
  }
}
function isMiniappMode() {
  return typeof window !== "undefined" && window.parent !== window;
}
function onAppData(fn) {
  _dataListeners.push(fn);
}
function onWalletChange(fn) {
  _listeners.push(fn);
  fn(_address);
  return () => {
    const idx = _listeners.indexOf(fn);
    if (idx !== -1)
      _listeners.splice(idx, 1);
  };
}
function sendTransactions(transactions) {
  return new Promise((resolve, reject) => {
    const requestId = "req_" + ++_requestCounter;
    _pending[requestId] = { resolve, reject };
    window.parent.postMessage({ type: "send_transactions", requestId, transactions }, "*");
  });
}
function signMessage(message, signatureType = "erc1271") {
  return new Promise((resolve, reject) => {
    const requestId = "req_" + ++_requestCounter;
    _pending[requestId] = { resolve, reject };
    window.parent.postMessage({ type: "sign_message", requestId, message, signatureType }, "*");
  });
}
export {
  signMessage,
  sendTransactions,
  onWalletChange,
  onAppData,
  isMiniappMode
};
